File: Arrays/groupAnagrams.py
class Solutio1:
    def groupAnagrams(self, strs):
        # {sorted_word: [group]}
        # each key is a unique sorted_word
        hash_map = {}


        sorted_strs = []   # stores strings in strs but sorted
        for word in strs:  
            sorted_strs.append("".join(sorted(word)))    # add sorted-str into list

        # iterate sorted strings
        for word in sorted_strs:
            if word not in hash_map.keys(): # if it doesnt exist in map, add it with empty list value
                hash_map[word] = []

        # iterate orginal strings in strs
        for word in strs:
            sorted_str = "".join(sorted(word))  # sort cur-string
            for key in hash_map.keys(): # iterate each key in map, each unqiue sorted word
                if sorted_str == key:  # if cur-string-sorted in equal to unique-sorted-word-key
                    hash_map[key].append(word)  # add orginal cur-word to the unique-sorted-word-key's values list

        output = [] # each element is a list group of anagrams
        for key in hash_map.keys():   # iterate unique-sorted-keys and get each group of anagrams as list
            output.append(hash_map[key])    # and add it to output-array
        return output
    

class Solution2:
    def groupAnagrams(self, strs):

        storage = {} # {unqiue: indx}
        output = [] # [[group1], [group2], ....]

        # iterate each string in strs
        for s in strs:
            sorted_s =  "".join(sorted(s))  # sort current-string 
            # if sorted-cur-string is not is storage-hash-map
            if sorted_s not in storage.keys(): 
                output.append([])   # add an empty list for output-array for it
                # add sorted-string to map, set index of group to be current last index of output-list
                storage[sorted_s] = len(output)-1  
            # if sorted-strin exists in storage then we can add it to its gorup
            if sorted_s in storage:
                indx = storage[sorted_s] # get index of group using sorted-string
                output[indx].append(s)  # add orignal-current-unsorted-string to the indx of sorted-string in output-list

        return output

File: Arrays/GroupAnagrams.cpp
// TIME LIMIT EXCEEDED
class Solution {
public:

    vector<vector<string>> groupAnagrams(vector<string>& strs) {
        vector<vector<string>> output;
        if (strs.size() == 0) { 
            return output;
        }

        // iterate strings, and start creating groups
        for (int i = 0; i < strs.size(); ++i) { 
            output = addOrCreateNewGroup(output, strs[i]);
        }
        
        return output;
    }

    // adds a string to a group or creates a new group
    vector<vector<string>> addOrCreateNewGroup(vector<vector<string>> output, string base) {
        
        // if it fits in a group then add it to that group and return.
        // if it doesnt fit in the group, then create a new group with only it in it. return
        // iterate all groups in output
        bool foundAgroup = false;
        for (int i = 0; i < output.size(); ++i) { 
            // if given string is anagram in cur-group, add it to the cur-group
            if (isAnagram(base, output[i][0]) == true) {
                output[i].push_back(base);
                foundAgroup = true;
            }
        }
        if (foundAgroup == false) {
            vector<string> new_group;
            new_group.push_back(base);
            output.push_back(new_group);
        }
        return output;
    }

    bool static isAnagram(string base, string other) {
        // if they are sasme string they are anagram
        if (base == other) {
            return true;
        }
        // since empty string is not anagram of "b", it is not anagram if one of them is a empty return false
        if (base.empty() != other.empty()) {
            return false;
        }
        // if they are not same lengths they cannot be anagrams
        if (base.size() != other.size()) {
            return false;
        }

        // count freuencies of each char in first-word
        map<char, int> count1;
        for (char cur_char : base) {
            // if cur-char doesnt already exist then it init with key/val w 0, and adds 1, else just increments
            count1[cur_char] += 1;
        }

        // get counts for second word
        map<char, int> count2;
        for (char cur_char : other) {
            count2[cur_char] += 1;
        }

        // iterate every key-val pair in map, .first=key, .second=value
        for (const auto& pair : count1) {
            // if the counts of the char in both words are not equal then no anagrams
            if (count2[pair.first] != pair.second) {
                return false; 
            }
        }

        return true;
    }
};